sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,577 @@
+  final val WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER_END = 'R'
+
+  // This class represents one or more characters that we expect to be present in the input string
+  // based on the format string.
+  abstract class InputToken()
+  // Represents some number of digits (0-9).
+  abstract class Digits extends InputToken
+  // Represents exactly 'num' digits (0-9).
+  case class ExactlyAsManyDigits(num: Int) extends Digits
+  // Represents at most 'num' digits (0-9).
+  case class AtMostAsManyDigits(num: Int) extends Digits
+  // Represents one decimal point (.).
+  case class DecimalPoint() extends InputToken
+  // Represents one thousands separator (,).
+  case class ThousandsSeparator() extends InputToken
+  // Represents one or more groups of Digits (0-9) with ThousandsSeparators (,) between each group.
+  // The 'tokens' are the Digits and ThousandsSeparators in order; the 'digits' are just the Digits.
+  case class DigitGroups(tokens: Seq[InputToken], digits: Seq[Digits]) extends InputToken
+  // Represents one dollar sign ($).
+  case class DollarSign() extends InputToken
+  // Represents one optional plus sign (+) or minus sign (-).
+  case class OptionalPlusOrMinusSign() extends InputToken
+  // Represents one optional minus sign (-).
+  case class OptionalMinusSign() extends InputToken
+  // Represents one opening angle bracket (<).
+  case class OpeningAngleBracket() extends InputToken
+  // Represents one closing angle bracket (>).
+  case class ClosingAngleBracket() extends InputToken
+  // Represents any unrecognized character other than the above.
+  case class InvalidUnrecognizedCharacter(char: Char) extends InputToken
+}
+
+/**
+ * This class represents a parser to implement the to_number SQL function.
+ *
+ * It works by consuming an input string and a format string. This class accepts the format string
+ * as a field, and proceeds to iterate through the format string to generate a sequence of tokens
+ * (or throw an exception if the format string is invalid). Then when the function is called with an
+ * input string, this class steps through the sequence of tokens and compares them against the input
+ * string, returning a Spark Decimal object if they match (or throwing an exception otherwise).
+ *
+ * @param originNumberFormat the format string describing the expected inputs.
+ * @param errorOnFail true if evaluation should throw an exception if the input string fails to
+ *                    match the format string. Otherwise, returns NULL instead.
+ */
+class ToNumberParser(originNumberFormat: String, errorOnFail: Boolean) extends Serializable {
+  import ToNumberParser._
+
+  // Consumes the format string and produce a sequence of input tokens expected from each input
+  // string.
+  private lazy val formatTokens: Seq[InputToken] = {
+    val tokens = mutable.Buffer.empty[InputToken]
+    var i = 0
+    var reachedDecimalPoint = false
+    val format = originNumberFormat
```

Comment:
nit: seems we can just rename `originNumberFormat` to `format`

Reply:
Done.